crates/storage/nippy-jar/src/writer.rs

```diff
@@ -0,0 +1,401 @@
+            if length > 1 {
+                // first byte is reserved for `bytes_per_offset`, which is 8 initially.
+                let num_offsets = (length - 1) / OFFSET_SIZE_BYTES;
+
+                if remaining_to_prune as u64 > num_offsets {
+                    return Err(NippyJarError::InvalidPruning(
+                        num_offsets,
+                        remaining_to_prune as u64,
+                    ))
+                }
+
+                let new_num_offsets = num_offsets.saturating_sub(remaining_to_prune as u64);
+
+                // If all rows are to be pruned
+                if new_num_offsets <= 1 {
+                    // <= 1 because the one offset would actually be the expected file data size
+                    self.offsets_file.set_len(1)?;
+                    self.data_file.set_len(0)?;
+                } else {
+                    // Calculate the new length for the on-disk offset list
+                    let new_len = 1 + new_num_offsets * OFFSET_SIZE_BYTES;
+                    // Seek to the position of the last offset
+                    self.offsets_file
+                        .seek(SeekFrom::Start(new_len.saturating_sub(OFFSET_SIZE_BYTES)))?;
+                    // Read the last offset value
+                    let mut last_offset = [0u8; OFFSET_SIZE_BYTES as usize];
+                    self.offsets_file.read_exact(&mut last_offset)?;
+                    let last_offset = u64::from_le_bytes(last_offset);
+
+                    // Update the lengths of both the offsets and data files
+                    self.offsets_file.set_len(new_len)?;
+                    self.data_file.set_len(last_offset)?;
+                }
+            } else {
+                return Err(NippyJarError::InvalidPruning(0, remaining_to_prune as u64))
+            }
+        }
+
+        self.offsets_file.seek(SeekFrom::End(0))?;
+        self.data_file.seek(SeekFrom::End(0))?;
+
+        self.jar.rows = self.jar.rows.saturating_sub(num_rows);
+        if self.jar.rows == 0 {
+            self.jar.max_row_size = 0;
+        }
+        self.jar.freeze_config()?;
+
+        Ok(())
+    }
+
+    /// Updates [`NippyJar`] with the new row count and maximum uncompressed row size, while
+    /// resetting internal fields.
+    fn finalize_row(&mut self) {
+        self.jar.max_row_size = self.jar.max_row_size.max(self.uncompressed_row_size);
+        self.jar.rows += 1;
+
+        self.tmp_buf.clear();
+        self.uncompressed_row_size = 0;
+        self.column = 0;
+    }
+
+    /// Commits configuration and offsets to disk. It drains the internal offset list.
+    pub fn commit(&mut self) -> Result<(), NippyJarError> {
+        self.commit_offsets()?;
+
+        // Flushes `max_row_size` and total `rows` to disk.
+        self.jar.freeze_config()?;
+
+        Ok(())
+    }
+
+    /// Flushes offsets to disk.
+    pub(crate) fn commit_offsets(&mut self) -> Result<(), NippyJarError> {
```

Comment:
this should probably do `sync_all` to attempt to tell the OS to actually flush the data to disk and not just its i/o buffer https://doc.rust-lang.org/std/fs/struct.File.html#method.sync_all